### PR TITLE
relation ships are now saved after the record itself

### PR DESCRIPTION
**Pull request by olivierchatry**
solve #327, save record and then save relationships, which allows for rules to be applied ( generally rules are using record data, not relationship data).

### CODE REVIEW

**Comment by olivierchatry** [comment]
I signed it!

**Comment by olivierchatry** [comment]
Seems like the last commit did not trigger a rebuild :(, @tstirrat any chance you can help or should I just push a dummy commit ? 

I also found why I needed to reload some records after saving it happens when you create a record with relationship belongsTo -> hasMany relationship, and save the belongsTo ( parent ) first. Ember data has some bad code that check if the relationship element is "new" and is so push it again in the "hasMany" relationship. I have filled a bug at ember data, but you can fix it by saving the child before the parent ( in some case it it is not possible ).

**Comment by tstirrat** [comment]
for reference: emberjs/data#4149

**Comment by tstirrat** [comment]
I tried to kick off the build again but it runs on the second to last commit, sorry

**Comment by olivierchatry** [comment]
Submitted à pullrequest to ember data for the fix as well as test for repro & proof of fix

**Comment by olivierchatry** [comment]
Then I'm going to submit a dummy

File: addon/adapters/firebase.js [hunk 1]
```diff
@@ -418,42 +418,81 @@ export default DS.Adapter.extend(Waitable, {
...
-          rejected.push(promises.record);
-        }
-        // Throw an error if any of the relationships failed to save
-        if (rejected.length !== 0) {
-          var error = new Error(`Some errors were encountered while saving ${typeClass} ${snapshot.id}`);
-          error.errors = Ember.A(rejected).mapBy('reason');
-          reject(error);
-        } else {
-          resolve();
+      recordPromise.then(
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: you can combine this with the prior line `this._updateRecord().then()`

File: addon/adapters/firebase.js [hunk 2]
```diff
@@ -418,42 +418,81 @@ export default DS.Adapter.extend(Waitable, {
...
-            delete serializedRecord[key];
-          }
-        } else {
-          if (this.isRelationshipEmbedded(store, typeClass.modelName, relationship) && serializedRecord[key]) {
-            save = this._saveEmbeddedBelongsToRecord(store, typeClass, relationship, serializedRecord[key], recordRef);
-            savedRelationships.push(save);
+        const data = serializedRecord[key];
+        const isEmbedded = this.isRelationshipEmbedded(store, typeClass.modelName, relationship);
+        const hasMany = relationship.kind === 'hasMany';
+        if ( hasMany || isEmbedded ) {
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: remove the padding inside the `()`, to be consistent with the code style elsewhere

File: addon/adapters/firebase.js (hunk 1 above)
Comment (olivierchatry) [inline addon/adapters/firebase.js]: I dont think I can, since recordPromise is used later in the callback

File: addon/adapters/firebase.js [hunk 3]
```diff
@@ -418,42 +418,81 @@ export default DS.Adapter.extend(Waitable, {
...
-          resolve();
+      recordPromise.then(
+        () => {
+          if (recordPromise.state === 'rejected') {
+            // TODO: FIXME: Do we need to rollback ? I think we do.
+            reportError([recordPromise.reason]);
+          } else {
+            // and now we construct the list of promise to save relationships.
+            var savedRelationships = Ember.A();
+            relationshipsToSave.forEach(
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: this could be simplified with a map:

```js
var savedRelationships = relationshipsToSave.map((rel) => {})
```

File: addon/adapters/firebase.js (hunk 1 above)
Comment (tstirrat) [inline addon/adapters/firebase.js]: `recordPromise.state` should not be `rejected` after the promise resolved, it would be inside the rejected (`.catch()`)

Also, we would not need to rollback inside the `then` block. The good thing about your new approach, due to promise chaining, is that when Firebase returns an error, we will not attempt to save the relationships.

**Comment by olivierchatry** [comment]
@tstirrat any more comment ?

File: addon/adapters/firebase.js [hunk 4]
```diff
@@ -418,42 +418,81 @@ export default DS.Adapter.extend(Waitable, {
...
-        // Throw an error if any of the relationships failed to save
-        if (rejected.length !== 0) {
-          var error = new Error(`Some errors were encountered while saving ${typeClass} ${snapshot.id}`);
-          error.errors = Ember.A(rejected).mapBy('reason');
-          reject(error);
-        } else {
-          resolve();
+      recordPromise.then(
+        () => {
+          if (recordPromise.state === 'rejected') {
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: this check should not be here, it should be in the `.catch()` handler

Comment (olivierchatry) [inline addon/adapters/firebase.js]: Ok, but what about if the attributes are saving OK, but the relations are
failing ?

File: addon/adapters/firebase.js [hunk 5]
```diff
@@ -418,42 +418,81 @@ export default DS.Adapter.extend(Waitable, {
...
+            // TODO: FIXME: Do we need to rollback ? I think we do.
+            reportError([recordPromise.reason]);
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: The error handling code can be simplified to:

```js
this._updateRecord()
  .then(() => { /*... */ })
  .catch(reportError);
```

File: addon/adapters/firebase.js (hunk 4 above)
Comment (tstirrat) [inline addon/adapters/firebase.js]: this would probably be related to security rules. its a good question.

If we want to support this, we'd need a backup of previous attributes. This could be grabbed from ember-data.

Comment (olivierchatry) [inline addon/adapters/firebase.js]: Done, test are passing, but I did not try yet on my application ( waiting
for it to build ).

**Comment by tstirrat** [comment]
Thanks for this PR, btw. I appreciate all the help I can get!

One final nitpick: arrow functions blocks should be on the same line:

```js
method.then(() => {
  // ...
});
```

**Comment by olivierchatry** [comment]
no problem, I just dont like to wait for things to happen :). If my PR is validated for ember-data, that would be cool for emberfire as well ( and probably other adapter/serializer ).

File: addon/adapters/firebase.js [hunk 6]
```diff
@@ -418,42 +418,70 @@ export default DS.Adapter.extend(Waitable, {
...
-          reject(error);
-        } else {
-          resolve();
+      ).catch(
+        (e) => {
+          reportError([e]);
         }
-      });
+      ).then(
+        (results) => {
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: still a lambda here on new line

File: addon/adapters/firebase.js [hunk 7]
```diff
@@ -418,42 +418,70 @@ export default DS.Adapter.extend(Waitable, {
...
         }
-        // Throw an error if any of the relationships failed to save
-        if (rejected.length !== 0) {
-          var error = new Error(`Some errors were encountered while saving ${typeClass} ${snapshot.id}`);
-          error.errors = Ember.A(rejected).mapBy('reason');
-          reject(error);
-        } else {
-          resolve();
+      ).catch(
+        (e) => {
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: and another lambda here

**Comment by tstirrat** [comment]
so. we could reduce the number of saves to exactly 2 by using a multi-path update for the relationships

writing something like this:

```js
rootReference.child('posts').child(postId).update({
  // ignore unchanged comment links
  'comments/comment_5': null, // deleted
  'comments/comment_6': true, // added
  'comments/comment_7': true, // added
});
```

**Comment by olivierchatry** [comment]
yes  I think it should be easily feasible, the code is pretty much already
there. that would also greatly reduce the risk of errors

**Comment by tstirrat** [comment]
Actually, we could probably reduce it to 1, with an update like this:

```js
rootReference.child('posts').update({
  // model attributes:
  'post_1/title': 'Title',
  'post_1/body': 'Body here',

  // relationships:
  // ignore unchanged comment links
  'post_1/comments/comment_5': null, // deleted
  'post_1/comments/comment_6': true, // added
  'post_1/comments/comment_7': true, // added
});
```

**Comment by olivierchatry** [comment]
yep that make sense ( trying to figure out if something is wrong with either my app or the change I commited )

**Comment by tstirrat** [comment]
if you find out its an edge case, or a slight bug.. please write a test that exposes it

**Comment by olivierchatry** [comment]
it was working before the change I have made ( catch, etc. ) so I'm
guessing something is not working as it should.

**Comment by tstirrat** [comment]
The more I think about this, the more I wonder if we can take the other atomic save branch, take the serialized json hash, convert it to a multi-path update with only changed attributes.. and we solve both problems.

e.g.

if the serializer returns only changed attributes:

```js
{
  title: 'New title',
  comments: {
    comment_1: null, // deleted
    comment_5: true, // added
  }
}
```

we can convert this to a multi-path update:

```js
{
  'post_1/title': 'New title',
  'post_1/comment_1': null, // deleted
  'post_1/comment_5': true, // added
}
```

**Comment by olivierchatry** [comment]
Well that would definitly solve the problem :) can you try ?

**Comment by tstirrat** [comment]
However this will cause problems with validation rules where say, "post must have a body".

So, an adapted version of this is that we could save all model attributes, and only the changed relationships..

**Comment by olivierchatry** [comment]
yes, attributes are fine to overrides normally

**Comment by tstirrat** [comment]
I'll get better time to work on this over the weekend. Sadly wont get a chance today/tomorrow.

**Comment by olivierchatry** [comment]
I can try to do it then, as I think it would be a very cool way of doing :)

**Comment by olivierchatry** [comment]
Got a first version that works, but does not handle embedded record, going
to submit inside my fork, but your branch

**Comment by olivierchatry** [comment]
For some odd reason, I cannot update your branch to my fork without getting
conflict ( weird as I did not do anything on it ), I'm always 33 commit
behinds :(. Any chance you would know anything about that ?

**Comment by tstirrat** [comment]
ah, sorry. I rebased it, so your history and the remote one are out of sync.

delete the branch locally and then check it out again

```
git checkout master
git branch -D serialize-relationships-atomic
git checkout serialize-relationships-atomic
```

**Comment by olivierchatry** [comment]
Nop, not working, I always endup with conflict one way or the other. I will
solve the conflict by taking what is last on the branch I guess.

**Comment by olivierchatry** [comment]
I think, if you think this branch is OK ( it looks and works OK at least in my software ), we should merge it and release a version before the multipath update is working.

**Comment by olivierchatry** [comment]
squashed everything together.

File: addon/adapters/firebase.js [hunk 8]
```diff
@@ -418,38 +418,58 @@ export default DS.Adapter.extend(Waitable, {
...
+      };
       var recordPromise = this._updateRecord(recordRef, serializedRecord);
-
-      Ember.RSVP.hashSettled({relationships: relationshipsPromise, record: recordPromise}).then((promises) => {
-        var rejected = Ember.A(promises.relationships.value).filterBy('state', 'rejected');
-        if (promises.record.state === 'rejected') {
-          rejected.push(promises.record);
-        }
-        // Throw an error if any of the relationships failed to save
+      recordPromise.then(() => {
```
Comment (tstirrat) [inline addon/adapters/firebase.js]: We don't need recordPromise separated into a named variable anymore. We can just use `this._updateRecord().then()` now.

Otherwise, we're good to go on this PR.

**Comment by olivierchatry** [comment]
done :) if you have time, take a look at https://github.com/firebase/emberfire/pull/366 :)
